### PR TITLE
Changed the "fullscreen" config variable to an enum

### PR DESCRIPTION
**Pull request by Berulacks**
...Which can be either "Fullscreen", "FullscreenDesktop", or "Windowed".
I also udated the win32gl adapter to reflect this.
Also cleaned up some cruft in sdl.d and changed the two deprecated
logFatal() statements in linux.d to fatal()s.

### CODE REVIEW

**Comment by PxlBuzzard** [comment]
LGTM :+1:

**Comment by dislogical** [comment]
The difference between Fullscreen and FullscreenDesktop isn't immediately apparent. Could we do Fullscreen and FullscreenWindowed instead? That seems to be the accepted standard in games at this point.

File: source/dash/graphics/adapters/adapter.d [hunk 1]
```diff
@@ -16,7 +16,8 @@ abstract class Adapter
 private:
     uint _width, _screenWidth;
     uint _height, _screenHeight;
-    bool _fullscreen, _backfaceCulling, _vsync;
+    bool _backfaceCulling, _vsync;
+    WindowType _mainWindowType;
```
Comment (dislogical) [inline source/dash/graphics/adapters/adapter.d]: Why `_mainWindowType`? Why not just `_windowType`?

File: source/dash/graphics/adapters/sdl.d [hunk 2]
```diff
@@ -103,7 +92,24 @@ public:
...
+
+        switch(config.display.windowMode)
+        {
+            case WindowType.Fullscreen:
+                trace("Setting main window to Fullscreen mode!");
+                window.setFullscreenSetting( SDL_WINDOW_FULLSCREEN );
+                break;
+            case WindowType.FullscreenDesktop:
+                trace("Setting main window to FullscreenDesktop mode!");
+                window.setFullscreenSetting( SDL_WINDOW_FULLSCREEN_DESKTOP );
```
Comment (dislogical) [inline source/dash/graphics/adapters/sdl.d]: Based on [the SDL docs](https://wiki.libsdl.org/SDL_WindowFlags), it seems like all this does is just use the default desktop resolution. It seems to me that we want `SDL_WINDOW_BORDERLESS` at the native resolution instead. That is, assuming we rename this to `FullscreenWindowed`.

File: source/dash/graphics/adapters/win32gl.d [hunk 3]
```diff
@@ -225,7 +225,7 @@ public:
             style |= DWS_WINDOWED;
         }
 
-        if( _wasFullscreen != fullscreen )
+        if( _wasFullscreen != ( mainWindowType == WindowType.Fullscreen) )
```
Comment (dislogical) [inline source/dash/graphics/adapters/win32gl.d]: Spacing here is weird? Looks like you're missing a space after `Fullscreen`.

File: source/dash/graphics/adapters/win32gl.d [hunk 4]
```diff
@@ -234,7 +234,7 @@ public:
                           SWP_NOZORDER | SWP_NOACTIVATE | SWP_FRAMECHANGED );
         }
 
-        _wasFullscreen = fullscreen;
+        _wasFullscreen = ( mainWindowType == WindowType.Fullscreen);
```
Comment (dislogical) [inline source/dash/graphics/adapters/win32gl.d]: Ditto.

File: source/dash/graphics/adapters/win32gl.d [hunk 5]
```diff
@@ -281,7 +281,7 @@ public:
     */
     final void openWindow( bool showWindow )
     {
-        hWnd = CreateWindowEx( 0, DGame.instance.title.ptr, DGame.instance.title.ptr, fullscreen ? DWS_FULLSCREEN : DWS_WINDOWED,
+        hWnd = CreateWindowEx( 0, DGame.instance.title.ptr, DGame.instance.title.ptr, (mainWindowType == WindowType.Fullscreen ) ? DWS_FULLSCREEN : DWS_WINDOWED,
```
Comment (dislogical) [inline source/dash/graphics/adapters/win32gl.d]: Missing a space before `mainWindowType`.

File: source/dash/utility/config.d [hunk 6]
```diff
@@ -17,6 +17,15 @@ const(Config) config() @property
     return configInst;
 }
 
+
```
Comment (dislogical) [inline source/dash/utility/config.d]: Too many blank lines here.

File: source/dash/utility/config.d [hunk 7]
```diff
@@ -61,12 +70,13 @@ public:
...
     {
-        @rename( "Fullscreen" )
-        bool fullscreen;
+        @rename( "WindowType" ) @optional @byName
+        WindowType windowMode = WindowType.Windowed;
         @rename( "Height" ) @optional
         uint height = 1920;
         @rename( "Width" ) @optional
         uint width = 720;
+
```
Comment (dislogical) [inline source/dash/utility/config.d]: Shouldn't be a blank line here.

**Comment by Berulacks** [comment]
> Based on the SDL docs, it seems like all this does is just use the default desktop resolution. It seems to me that we want SDL_WINDOW_BORDERLESS at the native resolution instead. That is, assuming we rename this to FullscreenWindowed.

Aha. Good point. I'll add a Borderless Window option.

> Why _mainWindowType? Why not just _windowType?

A variable called _windowType of type WindowType felt a bit off to me. Just semantics, really. If you prefer _windowType I'll switch it to that.

**Comment by Berulacks** [comment]
> Shouldn't be a blank line here.

I regret everything I've ever done.

**Comment by dislogical** [comment]
I would prefer `_windowType`, "main" doesn't mean or add anything, and just makes it longer.

Just so you know, you can directly reply to the line notes.

**Comment by Berulacks** [comment]
> Just so you know, you can directly reply to the line notes.

:+1:  Cheers, totally missed that.

**Comment by Berulacks** [comment]
Should be good, now.
http://youtu.be/GsjvXnjkTLA

**Comment by PxlBuzzard** [comment]
@Berulacks You uploaded that video just for this comment? That's some commitment.

**Comment by Berulacks** [comment]
No one on the Internet had uploaded that clip, yet (with relatively decent quality). I was feeling productive.

**Comment by dislogical** [comment]
I still don't think we need `FullscreenDesktop`. Seems like we should just merge it with `BorderlessWindow` to make `FullscreenWindowed`, that's just a borderless window, but add native resolution.

**Comment by Berulacks** [comment]
## Do we really want that, though? I mean, think about it: the full screen desktop "feature" exists, I'm assuming, as a convenience, in SDL. If that's the case, then wouldn't full-screen windowed fall under the same category of a minor convenience? I feel like it might be best to either include both options, or neither. Though, I'm probably thinking too much into this. I'll make a commit when my plane lands. 

-Beru

**Comment by dislogical** [comment]
I mean, isn't `FULLSCREEN_DESKTOP` just the same as `FULLSCREEN` at the window resolution? If so, it's probably not worth adding a whole option for it, especially one we have to maintain. My vote: Remove `FullscreenDesktop`, rename `BorderlessWindowed` to `FullscreenWindowed`, and make it adopt the screen resolution regardless of config settings.

**Comment by Berulacks** [comment]
Wow. Totally forgot about this, my bad. I'm removing FullscreenDesktop and creating a FullscreenWindowed. Are we sure we don't want to keep a borderless windowed option, though? Seems like something some people might want it.

**Comment by dislogical** [comment]
I do like the idea of BorderlessWindow, it was mostly the FullscreenDesktop thing I thought seemed redundant.

Nice work, btw. Currently in the Chicago airport, but I should be able to do a final-ish review on Tuesday or Wednesday.

**Comment by Berulacks** [comment]
:+1: :+1:

File: source/dash/graphics/adapters/sdl.d [hunk 8]
```diff
@@ -103,7 +92,35 @@ public:
...
+            case WindowType.Fullscreen:
+                trace("Setting main window to Fullscreen mode!");
+                window.setFullscreenSetting( SDL_WINDOW_FULLSCREEN );
+                break;
+            case WindowType.FullscreenWindowed:
+                trace("Setting main window to Fullscreen Windowed mode!");
+                bordered = false;
+                width = screenWidth;
+                height = screenHeight;
+                goto default;
```
Comment (dislogical) [inline source/dash/graphics/adapters/sdl.d:115]: I like that this is a reasonable use of `goto`.

File: source/dash/graphics/adapters/sdl.d [hunk 9]
```diff
@@ -103,7 +92,35 @@ public:
...
+                bordered = true;
+                goto default;
+            case WindowType.BorderlessWindow:
+                trace("Setting main window to Borderless Windowed mode!");
+                bordered = false;
+                goto default;
+            default: 
+                window.setFullscreenSetting( 0 );
+                window.setSize( width, height );
+                SDL_SetWindowBordered( SDL_GetWindowFromID( window.id() ), bordered );
```
Comment (dislogical) [inline source/dash/graphics/adapters/sdl.d]: This looks pretty grody. Is there any way to get a `window.setWindowBordered` (or similar) into `gfm`?

**Comment by dislogical** [comment]
Aside from the `setWindowBordered` thing, LTGM :+1:

**Comment by Berulacks** [comment]
## I actually already sent them a pull request, and it was accepted. I'm heading home right now, when I get there I'll update the gfm dependency and change that. 

-Beru

**Comment by dislogical** [comment]
Looks good! Also looks like you've got conflicts... Could you rebase, and we'll merge? While you're at it, if you wanted to squash a few of those commits it would be highly appreciated.

**Comment by Berulacks** [comment]
Squashed a few commits, how's it look now? I can squash any others, if need be. Sorry for taking so long... this whole squashing business was rather new to me :)

**Comment by dislogical** [comment]
I'm confused... there are 12 commits now? Not sure how that happened. I'll see if I can clean it up.

**Comment by Berulacks** [comment]
## Haha. Man, I'm bad at this rebasing business. I'll try and figure out what I did wrong when I get out of this here class. I guess when I squashed the commits git duplicated them, somehow? My mistake. 

-Beru

**Comment by dislogical** [comment]
Alright, no worries. Ping us when it looks how you want it to look. Also worth noting you may have to do a force push for GitHub to take it.

**Comment by Berulacks** [comment]
Rebasing/squashing, it seems, makes it so that the latest two commits by you guys seem to get added to the PR? It's odd. Though, the duplication happened because I did a "git pull" after rebasing, then pushing. How does it look, now? Do you think the fact that a55726e and 62ee2a4 are part of the PR would cause issues?

**Comment by dislogical** [comment]
I think it would be ok, but I think that that's probably a sign that something else isn't right. Are you rebasing on top of `develop`?

**Comment by Berulacks** [comment]
I was rebasing on top of feature/window_type

**Comment by dislogical** [comment]
Oooh, you should be rebasing feature/window_type on top of master. At this point, let's just merge it.

**Comment by Berulacks** [comment]
Man, I'm such a scrub. Oh well. Back to the git manual.

**Pull request merged**